rcl_action/src/rcl_action/action_server.c

```diff
@@ -0,0 +1,759 @@
+      if (RCL_RET_OK != ret) {
+        ret_final = RCL_RET_ERROR;
+      }
+      // Fill in any gaps left in the array with pointers from the end
+      action_server->impl->goal_handles[i] = action_server->impl->goal_handles[num_goal_handles];
+      action_server->impl->goal_handles[num_goal_handles--] = NULL;
+
+      ++(*num_expired);
+    }
+  }
+
+  // Shrink goal handle array if some goals expired
+  if (*num_expired > 0u) {
+    void * tmp_ptr = allocator.reallocate(
+      action_server->impl->goal_handles, num_goal_handles, allocator.state);
+    if (!tmp_ptr) {
+      RCL_SET_ERROR_MSG("failed to shrink size of goal handle array");
+      ret_final = RCL_RET_ERROR;
+    } else {
+      action_server->impl->goal_handles = (rcl_action_goal_handle_t **)tmp_ptr;
+      action_server->impl->num_goal_handles = num_goal_handles;
+    }
+  }
+  return ret_final;
+}
+
+rcl_ret_t
+rcl_action_take_cancel_request(
+  const rcl_action_server_t * action_server,
+  void * ros_cancel_request)
+{
+  TAKE_SERVICE_REQUEST(cancel);
+}
+
+rcl_ret_t
+rcl_action_process_cancel_request(
+  const rcl_action_server_t * action_server,
+  const rcl_action_cancel_request_t * cancel_request,
+  rcl_action_cancel_response_t * cancel_response)
+{
+  if (!rcl_action_server_is_valid(action_server)) {
+    return RCL_RET_ACTION_SERVER_INVALID;  // error already set
+  }
+  RCL_CHECK_ARGUMENT_FOR_NULL(cancel_request, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(cancel_response, RCL_RET_INVALID_ARGUMENT);
+
+  rcl_allocator_t allocator = action_server->impl->options.allocator;
+  const size_t total_num_goals = action_server->impl->num_goal_handles;
+
+  // Storage for pointers to active goals handles that will be transitioned to canceling
+  // Note, we need heap allocation for MSVC support
+  rcl_action_goal_handle_t ** goal_handles_to_cancel =
+    (rcl_action_goal_handle_t **)allocator.allocate(
+    sizeof(rcl_action_goal_handle_t *) * total_num_goals, allocator.state);
+  if (!goal_handles_to_cancel) {
+    RCL_SET_ERROR_MSG("allocation failed for temporary goal handle array");
+    return RCL_RET_BAD_ALLOC;
+  }
+  size_t num_goals_to_cancel = 0u;
+
+  // Request data
+  const rcl_action_goal_info_t * request_goal_info = &cancel_request->goal_info;
+  const uint8_t * request_uuid = request_goal_info->uuid;
+  int64_t request_nanosec = _goal_info_stamp_to_nanosec(request_goal_info);
+
+  // Determine how many goals should transition to canceling
+  if (!uuidcmpzero(request_uuid) && (0u == request_nanosec)) {
+    // UUID is not zero and timestamp is zero; cancel exactly one goal (if it exists)
+    rcl_action_goal_info_t goal_info = rcl_action_get_zero_initialized_goal_info();
+    rcl_action_goal_handle_t * goal_handle;
+    for (size_t i = 0; i < total_num_goals; ++i) {
+      goal_handle = action_server->impl->goal_handles[i];
+      rcl_ret_t ret = rcl_action_goal_handle_get_info(goal_handle, &goal_info);
+      assert(RCL_RET_OK == ret);
```

Comment:
IIUC assert is compiled out in release builds. Is this an `assert()` because its calling out an assumption that the above call will always succeed? Is it possible the above call could fail in normal use?

Reply:
I thought it should always succeed, but after thinking about it some more, this can fail if the goal handle is finalized outside of the action server's control.

Reply:
Addressed in 8f9041e